server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputePluginMeasuresStep.java

```diff
@@ -0,0 +1,197 @@
+  private final MetricsDefinition[] metricsDefinitions;
+  private final ProjectSettingsRepository settings;
+
+  public ComputePluginMeasuresStep(TreeRootHolder treeRootHolder, MetricRepository metricRepository, MeasureRepository measureRepository, MetricsDefinition[] metricsDefinitions,
+    ProjectSettingsRepository settings) {
+    this.treeRootHolder = treeRootHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+    this.metricsDefinitions = metricsDefinitions;
+    this.settings = settings;
+  }
+
+  @Override
+  public void execute() {
+    new NewMetricDefinitionsVisitor().visit(treeRootHolder.getRoot());
+  }
+
+  private class NewMetricDefinitionsVisitor extends DepthTraversalTypeAwareVisitor {
+    public NewMetricDefinitionsVisitor() {
+      super(FILE, PRE_ORDER);
+    }
+
+    @Override
+    public void visitAny(org.sonar.server.computation.component.Component component) {
+      MetricsDefinition.NewMetricContext newMetricContext = new MetricsDefinition.NewMetricContext();
+      for (MetricsDefinition metricsDefinition : metricsDefinitions) {
+        metricsDefinition.define(newMetricContext);
+      }
+
+      // TODO use DirectAcyclicGraph to execute measure computation in the right order
+      for (MetricsDefinition.Metric metric : newMetricContext.getMetrics()) {
+        MeasureComputerContextImpl measureComputerContext = new MeasureComputerContextImpl(component, metric);
+        metric.getComputer().getMeasureComputer().compute(measureComputerContext);
+      }
+    }
+  }
+
+  @Override
+  public String getDescription() {
+    return "Compute measures from plugin";
+  }
+
+  private class MeasureComputerContextImpl implements MetricsDefinition.MeasureComputerContext {
+
+    private final org.sonar.server.computation.component.Component component;
+    private final MetricsDefinition.Metric metricDefinition;
+
+    public MeasureComputerContextImpl(org.sonar.server.computation.component.Component component, MetricsDefinition.Metric metricDefinition) {
+      this.component = component;
+      this.metricDefinition = metricDefinition;
+    }
+
+    @Override
+    public Settings getSettings() {
+      return settings.getProjectSettings(component.getKey());
+    }
+
+    @Override
+    public Component getComponent() {
+      return new ComponentImpl(component);
+    }
+
+    @Override
+    public Measure getMeasure(String metricKey) {
+      // TODO Validate that this metric owns to the metric input metrics
+      Optional<org.sonar.server.computation.measure.Measure> measure = measureRepository.getRawMeasure(component, metricRepository.getByKey(metricKey));
+      if (measure.isPresent()) {
+        return new MeasureImpl(measure.get());
+      }
+      return null;
+    }
+
+    @Override
+    public List<Measure> getChildrenMeasures(String metricKey) {
+      // TODO Validate that this metric owns to the metric input metrics
+      return null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Return an empty collection instead of null. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1168)